modules/optimization/src/outputs/ExodusOptimizationSteady.C

```diff
@@ -0,0 +1,43 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ExodusOptimizationSteady.h"
+
+registerMooseObject("OptimizationApp", ExodusOptimizationSteady);
+
+InputParameters
+ExodusOptimizationSteady::validParams()
+{
+  // Get the base class parameters
+  InputParameters params = Exodus::validParams();
+
+  // Return the InputParameters
+  return params;
+}
+
+ExodusOptimizationSteady::ExodusOptimizationSteady(const InputParameters & parameters)
+  : Exodus(parameters), _steady_exec(dynamic_cast<Steady *>(_app.getExecutioner()))
+{
+  if (!_steady_exec)
+    mooseError(
+        "ExodusOptimizationSteady output can only be used with the SteadyAndAdjoint or Steady"
+        "executioners to output a per-optimization iteration solution.");
+}
+
+void
+ExodusOptimizationSteady::incrementFileCounter()
+{
+  _file_num = _steady_exec->getIterationNumberOutput();
+}
+
+Real
+ExodusOptimizationSteady::getTimeStepForOutput()
+{
+  return _steady_exec->getIterationNumberOutput();
+}
```

Comment:
this could just be the behavior of the Exodus output and you can delete that class

Reply:
in fact if someone overrode these methods in their executioner, they could be surprised not to see the right behavior in their Exodus output

Reply:
@dschwen @lindsayad this is where we need to decide if these two lines (that likely should be in the base class as well, since they rely on something a dev would have implemented in their executioner) are worth creating a new class for

I dont like naming outputs after Executioners.

Reply:
I guess to move this up to "Exodus" we would need to move up the "output step number" APIs to Executioner, as we don't really want to condition the call to the step number routine in Exodus to the type of the executioner. 

So I'm fine with this object for now.
I think the solution that keeps in line with our design principles would be to either rework the SetupInterface (execute_on) or create a new interface between Outputs and Executioner/Executors that handles the determination of the step number.

Reply:
The thing is that the majority of the time, except when a user wants to do per-solve-iteration debugging and in this optimization module use case, the exodus output file name has nothing to do with executioner data. In the majority of cases, the exodus output file name only depends on adaptivity data, which we divorced from the executioner block deliberately (although that syntax still exists). So if we wanted to tie Exodus output (or generically any output) to executioner data, we may need to make the executioner aware of more things, or the executioner will have to go through a chain of calls to get the adaptivity step, etc.